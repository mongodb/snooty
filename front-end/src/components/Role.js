import React, { Component } from 'react';

export default class Role extends Component {

  constructor() {
    super();
    this.base = `https://docs.mongodb.com/manual/reference`;
    this.roleDataTypes = {
      query: (term) => `${this.base}/operator/query/${term}/#op._S_${term}`,
      dbcommand: (term) => `${this.base}/command/${term}/#dbcmd.${term}`,
      method: (term) => `${this.base}/method/${term}/#${term}`
    };
    this.codeRoles = ['binary'];
  }

  roleRendering() {
    // normal link
    if (this.props.nodeData.name === 'doc') {
      return (
        <a href={ this.props.nodeData.target }>
          { this.props.nodeData.label.value }
        </a>
      )
    } 
    // roles with interaction
    else if (this.roleDataTypes[this.props.nodeData.name]) {
      const termModified = this.props.nodeData.target.replace('()', '').replace('$', '');
      const href = this.roleDataTypes[this.props.nodeData.name](termModified);
      return (
        <a href={ href } onMouseEnter={ (e) => { this.props.modal(e, href) } }>
          { this.props.nodeData.label }
        </a>
      )
    } 
    // binary case is unique (maybe others will be as well)
    else if (this.codeRoles.includes(this.props.nodeData.name)) {
      const termModified = this.props.nodeData.target.substr(this.props.nodeData.target.indexOf('.') + 1);
      const href = `${this.base}/program/${termModified}/#${this.props.nodeData.target.replace('~', '')}`
      return (
        <a href={ href } className="reference external" onMouseEnter={ (e) => { this.props.modal(e, href) } }>
          <code className="xref mongodb mongodb-binary docutils literal notranslate">
            <span className="pre">
              { termModified }
            </span>
          </code>
        </a>
      )
    } else {
      return <span>==Role not implemented: { this.props.nodeData.name } ==</span>
    }
  }

  render() {
    return this.roleRendering()
  }

}