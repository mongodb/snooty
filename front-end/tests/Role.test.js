import React from 'react';
import { shallow } from 'enzyme';
import Role from '../src/components/Role';

// data for this component
import mockDataDoc from './data/Role-doc.test.json';
import mockDataRef from './data/Role-ref.test.json';

it('renders correctly role "doc"', () => {
  const tree = shallow(<Role nodeData={mockDataDoc} />);
  expect(tree).toMatchSnapshot();
});

it('renders correctly role "ref"', () => {
  const tree = shallow(<Role nodeData={mockDataRef} />);
  expect(tree).toMatchSnapshot();
});
