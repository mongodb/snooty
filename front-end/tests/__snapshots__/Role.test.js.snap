// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly role "binary" 1`] = `
<a
  className="reference external"
  href="https://docs.mongodb.com/manual/reference/program/mongod.exe/#bin.mongod.exe"
>
  <code
    className="mongodb-binary xref mongodb docutils literal notranslate"
  >
    <span
      className="pre"
    >
      mongod.exe
    </span>
  </code>
</a>
`;

exports[`renders correctly role "doc" 1`] = `
<a
  className="reference external"
  href="/cloud/atlas"
>
  Create an Atlas Account and Cluster
</a>
`;

exports[`renders correctly role "manual" 1`] = `
<a
  className="reference external"
  href="/manual/introduction"
>
  The MongoDB Server Manual
</a>
`;

exports[`renders correctly role "program" 1`] = `
<strong
  className="program"
>
  mongoimport
</strong>
`;

exports[`renders correctly role "ref" 1`] = `
<a
  className="reference external"
  href="https://docs.mongodb.com/manual/installation/#mongodb-supported-platforms"
>
  <span
    className="xref std std-ref"
  >
    Supported Platforms
  </span>
</a>
`;

exports[`renders correctly role "term" 1`] = `
<a
  className="reference external"
  href="dbpath"
>
  data directory
</a>
`;
